app/app.go

```diff
@@ -602,10 +615,111 @@ func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
 // EndBlocker application updates every end block
 func (app *RegenApp) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
 	resp := app.mm.EndBlock(ctx, req)
+
+	// TODO: update with mainnet validatorset
+	// replace with
+	// validatorsPubkeys, err := GetMainnetValidatorSetPubkeys()
+	validatorsPubkeys, err := GetTestnetValidatorSetPubKeys()
+	if err != nil {
+		panic(err)
+	}
+
+	// before the patch height continue the current wrong behavior.
+
+	// at patch height
+	// for all validators in app state (app.StakingKeeper.GetAllValidators), create a ValidatorUpdate with the correct power
+	// for all remaining validators in []VoteInfo that are not in app state, set their voting power to zero
+	if ctx.BlockHeight() == PATCH_HEIGHT {
+		var updates []abci.ValidatorUpdate
+		validators := app.StakingKeeper.GetAllValidators(ctx)
+		inValidatorSet := map[string]bool{}
+
+		for _, validator := range validators {
+			tmProtoPk, err := validator.TmConsPublicKey()
+			if err != nil {
+				panic(err)
+			}
+
+			valCons, err := validator.GetConsAddr()
+			if err != nil {
+				panic(err)
+			}
+
+			if validator.Jailed {
+				for _, voteInfo := range app.votesInfo {
+
+					if sdk.ConsAddress(voteInfo.Validator.Address).Equals(valCons) {
+						updates = append(updates, abci.ValidatorUpdate{
+							PubKey: tmProtoPk,
+							Power:  validator.ConsensusPower(sdk.DefaultPowerReduction),
+						})
+
+						// We need to reset the counter & array so that the validator won't be immediately slashed for downtime after unjail.
+						signInfo, found := app.SlashingKeeper.GetValidatorSigningInfo(ctx, valCons)
+						if found {
+							signInfo.MissedBlocksCounter = 0
+							signInfo.IndexOffset = 0
+							app.SlashingKeeper.SetValidatorSigningInfo(ctx, valCons, signInfo)
+							missedBlocks := app.SlashingKeeper.GetValidatorMissedBlocks(ctx, valCons)
+							for _, mb := range missedBlocks {
+								app.SlashingKeeper.SetValidatorMissedBlockBitArray(ctx, valCons, mb.Index, false)
```

Comment:
This method `app.SlashingKeeper.clearValidatorMissedBlockBitArray` is private, so here I am resetting `ValidatorMissedBlockBitArray`.